Ignore whitespace between two successive EncodedWords

talanis85 opened the pull request:
Quoting RFC 2047 6.2:
  When displaying a particular header field that contains multiple
  'encoded-word's, any 'linear-white-space' that separates a pair of
  adjacent 'encoded-word's is ignored.

Fixes: https://github.com/purebred-mua/purebred-email/issues/95

talanis85 committed:
Ignore whitespace between two successive EncodedWords

Quoting RFC 2047 6.2:
  When displaying a particular header field that contains multiple
  'encoded-word's, any 'linear-white-space' that separates a pair of
  adjacent 'encoded-word's is ignored.

Fixes: https://github.com/purebred-mua/purebred-email/issues/95

frasertweedale commented:
There was a property test failure caused by `encodeEncodedWords` intercalating `" "` between successive encoded words:
```
  encodeEncodedWords roundtrip:                                 FAIL (0.51s)
      ✗ <interactive> failed at tests/EncodedWord.hs:35:61
        after 561 tests and 12 shrinks.
        shrink path: 561:c4B2cFe2K2
      
           ┏━━ tests/EncodedWord.hs ━━━
        32 ┃ prop_roundtrip :: Property
        33 ┃ prop_roundtrip = property $ do
        34 ┃   t <- forAll $ Gen.text (Range.linear 0 1000) Gen.unicode
           ┃   │ "\NUL \NUL"
        35 ┃   decodeEncodedWords defaultCharsets (encodeEncodedWords t) === t
           ┃   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
           ┃   │ ━━━ Failed (- lhs) (+ rhs) ━━━
           ┃   │ - "\NUL\NUL"
           ┃   │ + "\NUL \NUL"
```

The implementation:

```haskell
encodeEncodedWords :: T.Text -> B.ByteString
encodeEncodedWords t = maybe utf8 (const ew) (chooseEncodedWordEncoding utf8)
  where
    ew = B.intercalate " " . fmap encOrNot . B.split 32 {- SPACE -} $ utf8
    encOrNot s = maybe s (g s) (chooseEncodedWordEncoding s)
    g s (enc, p) = serialiseEncodedWord $
      EncodedWord "utf-8" Nothing enc (review (clonePrism p) s)
    utf8 = T.encodeUtf8 t
```

Just brain-dumping an idea on how to fix.  Feel free to resolve it or leave it for me (I'll get to it in the next week or so).

- instead of `fmap encOrNot`...
- first `fmap (\s -> (chooseEncodedWordEncoding s, s)`
- detect runs of multiple `Just _` (e.g. using `Data.List.NonEmpty.groupWith (isJust . fst)`)
- for each run of of `(Just _, s):|t`, cons a SPACE on each string in the tail
- serialise and intercalate

talanis85 committed:
Encode spaces in consecutive encoded words correctly.

When encodeEncodedWords produces two or more consecutive encoded words,
actually encode the space between, since it will be ignored on decoding.

talanis85 commented:
Added another commit to fix `encodeEncodedWords`. It now actually encodes spaces between two encoded words.
